### ISSUE TITLE
Timeline zoom limit and centering

### ISSUE THREAD

**Issue by RCarvalhoNASA** [nasa/openmct#936]
I'm working on a very big timeline in the stable version of WARP, and I'm finding it difficult to get in and make adjustments in part due to how the zoom works.  Zooming out has a maximum of only about 5 days on the screen (and this timeline would ideally be months long).  Zooming in to be able to get a handle on some of the smaller activities in there is nearly impossible, as the zoom in range does not bear much resemblence to where I'm looking right now.  Ideally, the zoom should remain centered on the same point that the current view is centered on, whether zooming in or out.  It seems to be rather arbitrary where the zoom puts the user.

**Comment by charlesh88** [comment]
@RCarvalhoNASA Talked to Robert about this today 5/23. This is impacting his ability to build timelines for extended mission timelines for RP for Nobile crater - we should look to fast track a fix for this. Robert's timeline is in the stable build, and it's named "Nobile timeline 2016". As of today, the content in there is about a quarter of what he wants to include. Robert, please add/edit as needed.

The solution here involves several elements:
* Zoom must be able to allow zooming out to view the full temporal extent of the Timeline.
* The Timeline should initially load displaying its meaningful full temporal extents. By this, we mean two things:
 * If the outermost Timeline begins at an offset greater than 0, don't include that offset - position and scale the Timeline so that its first nested child starts at the left side of the view area.
 * The end of the outermost Timeline would be in view at the right edge of the view.
* Zooming (both in and out) should be centered on the timestamp of whatever is centered in the current view area. In other words, if SET 1:20 is centered, 1:20 should remain centered after any zooming operations (both in or out) unless the operation would cause the left side of the Timeline's view area to become less than 0 while zooming out. At that point, it would be acceptable for the original center to proportionally shift as more of the temporal extents come into view.
* We might argue that the Timeline can't be zoomed out past its temporal extents, but I don't think that's a requirement.
* We discussed the drawbacks of centered zooming, such as cases where you want to zoom from the left, and ideas there were to allow a click to establish a Time of Interest point around which zoom actions would center. The TOI would always be visible as a vertical line, and the user could click somewhere TBD (perhaps in the ticks area) to set the TOI position. 
* We discussed mouse wheel zooming centered on cursor position, but Robert balked at this citing a belief this might be hard to control. 
* There is currently a minimum width issue with Timeline and Activity elements: they can't get small enough. This may be a CSS problem. Elements should be able to be scaled to a minimum width of 1 or 2 px. We may need to disallow click/drag actions unless an element has sufficient width TBD. @charlesh88 will look into this.

@VWoeltjen Putting this in the current sprint due to the nature of the need. We should discuss relative difficulty of the options here. View-centered zooming might be the easiest fix to implement quickly, followed by a TOI element. It still might be worth investigating mouse wheel zooming if we can plug in a library and test drive it without much effort.

**Comment by RCarvalhoNASA** [comment]
My other issue with the mousewheel zoom is that it would conflict with the scroll using the mousewheel.

**Comment by VWoeltjen** [comment]
> The Timeline should initially load displaying its meaningful full temporal extents.

We currently persist the zoom level for timelines that have been edited (so if you Edit and save, whatever zoom level you were last at will be the default zoom when a timeline is opened)

Should this behavior be replaced, or should this initial zoom apply only when no zoom level has been persisted?

Also, would it simplify anything if we had a "zoom to fit" button?

**Comment by RCarvalhoNASA** [comment]
The zoom to fit button would be excellent.  As for persisting the zoom, that seems like desirable behaviour when one is coming out of edit mode via a save.  However, if one is coming on the timeline fresh, starting in the zoom to fit state would be better.  If this is not possible, I would think the zoom to fit would be the desirable default.

**Comment by larkin** [comment]
I'd like to defer this pending time conductor overhaul and API updates. Timelines should have time synchronization too, which means extents should follow time conductor.  Additionally, Timelines should have a gesture that can be used to set the time conductor to the full extent of the timeline. Both of these cases are handled by time conductor overhaul. 

Rushing this over time conductor would be inefficient; we have to walk before we can run.

**Comment by larkin** [comment]
(If there is a quick fix for this, okay, but we shouldn't spend too much time)

**Comment by VWoeltjen** [comment]
Discussed at tag-up May 24; in terms of what to defer and what not to defer, will add centered zoom and zoom-to-fit behaviors (both on load and on button press)

**Comment by VWoeltjen** [comment]
Filed #955 to address a subset of the bullet points above. @charlesh88 - I have some questions about which, if any, of these points need to be tracked in follow up issues (to constrain the scope of this issue). A breakdown:

> * Zoom must be able to allow zooming out to view the full temporal extent of the Timeline.

Addressed in #955; there's still a max zoom-out, but it is large enough to see 2-3 years on a small display.

> * The Timeline should initially load displaying its meaningful full temporal extents. By this, we mean two things:
 * If the outermost Timeline begins at an offset greater than 0, don't include that offset - position and scale the Timeline so that its first nested child starts at the left side of the view area.
 * The end of the outermost Timeline would be in view at the right edge of the view.

Implemented in #955. As a consequence of using fixed intervals as zoom levels, there ends up being some margin at the right hand side in most cases.

Also removed persistence of zoom level to support this, per discussion above. (The one case where this would be useful is when leaving edit mode, but this case is not easy to support without broader changes to view lifecycle handling.)

> * Zooming (both in and out) should be centered on the timestamp of whatever is centered in the current view area. In other words, if SET 1:20 is centered, 1:20 should remain centered after any zooming operations (both in or out) unless the operation would cause the left side of the Timeline's view area to become less than 0 while zooming out. At that point, it would be acceptable for the original center to proportionally shift as more of the temporal extents come into view.

Implemented in #955

> * We might argue that the Timeline can't be zoomed out past its temporal extents, but I don't think that's a requirement.

Believe there's no action here; there's some utility to zooming out farther (gives you room for editing)

> * We discussed the drawbacks of centered zooming, such as cases where you want to zoom from the left, and ideas there were to allow a click to establish a Time of Interest point around which zoom actions would center. The TOI would always be visible as a vertical line, and the user could click somewhere TBD (perhaps in the ticks area) to set the TOI position. 

Have not acted on this since for scope, and because this may interact with time conductor. Should we file a separate issue?

> * We discussed mouse wheel zooming centered on cursor position, but Robert balked at this citing a belief this might be hard to control. 

We mentioned control keys as an option to avoid ambiguity with normal scroll bar control. Have not acted on this. Should we file a separate issue to implement this?

> * There is currently a minimum width issue with Timeline and Activity elements: they can't get small enough. This may be a CSS problem. Elements should be able to be scaled to a minimum width of 1 or 2 px. We may need to disallow click/drag actions unless an element has sufficient width TBD. @charlesh88 will look into this.

Have not acted on this. Should we file a separate issue?

**Comment by VWoeltjen** [comment]
Addressed by #955, at least to the extent described above. #971 tracks zooming with scroll wheel and #965 tracks minimum size of gantt bars.

**Comment by charlesh88** [comment]
Testathon 20160531
Verified fixed.

**Comment by charlesh88** [comment]
Testathon 2016-06-02
Timeline is not zooming on center as expected.

**Comment by VWoeltjen** [comment]
Seeing zoom-to-center issues as well. In some cases it works fine, in some cases you end up way earlier in the timeline than you should. Think this is same root cause as [regression on align-to-start behavior](https://github.com/nasa/openmct/issues/817#issuecomment-223434831).

**Comment by charlesh88** [comment]
Testathon 2016-06-03
Verified fixed, with caveat that (although I don't consider this a blocker) we really need to fix the Gantt area flashing problem while zooming, #997.
